poptorch_experimental_addons/_impl/collectives.py

```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import torch
+import poptorch
+
+
+def replicated_all_gather(x: torch.Tensor, replication_factor: int) -> torch.Tensor:
+    """
+    SPMD all-gather across IPUs.
+
+    Gathers and concatenates tensors occupying the same memory location across IPUs,
+    then replicates the result.
+    """
+    out = poptorch.custom_op(
+        [x],
+        name="ReplicatedAllGather",
+        domain="ai.graphcore",
+        domain_version=1,
+        example_outputs=[torch.randn(replication_factor, *x.shape)],
```

Comment:
Perhaps just use `zeros_like` or `full(nan)`. I feel a bit squeamish about modifying the global RNG state.

Reply:
Yes - oversight on my part